linera-core/src/chain_worker/state/mod.rs

```diff
@@ -309,26 +309,78 @@ where
     }
 
     /// Returns an error if the block requires a blob we don't have, or if unrelated blobs were provided.
-    async fn check_no_missing_blobs(
+    async fn check_no_missing_confirmed_blobs(
         &self,
-        blobs_in_block: HashSet<BlobId>,
+        mut required_blob_ids: HashSet<BlobId>,
+        blobs: &[Blob],
+    ) -> Result<(), WorkerError> {
+        self.check_for_unneeded_blobs_and_update_required(&mut required_blob_ids, blobs)
+            .await?;
+
+        let missing_blob_ids = self
+            .recent_blobs
+            .subtract_cached_items_from::<_, Vec<_>>(required_blob_ids, |id| id)
+            .await;
+
+        let missing_blob_ids = self.storage.missing_blobs(missing_blob_ids).await?;
+        if missing_blob_ids.is_empty() {
+            return Ok(());
+        }
+
+        Err(WorkerError::BlobsNotFound(missing_blob_ids))
+    }
+
+    /// Returns an error if the block requires a blob we don't have, or if unrelated blobs were provided.
+    async fn check_no_missing_validated_blobs(
+        &self,
+        mut required_blob_ids: HashSet<BlobId>,
         blobs: &[Blob],
     ) -> Result<(), WorkerError> {
-        let missing_blobs = self.get_missing_blobs(blobs_in_block, blobs).await?;
+        self.check_for_unneeded_blobs_and_update_required(&mut required_blob_ids, blobs)
+            .await?;
+        let pending_blobs = &self.chain.manager.get().pending_blobs;
```

Comment:
What's the current policy for pending blobs in the chain manager? Why should we check it when processing a validated block, but not a confirmed block?

Reply:
So right now we only add blobs to the chain manager when we create a validation vote. We currently don't remove them, but that will be fixed soon (#2704). So my naming here kind of sucks (I'll change it), but this means that we're checking for blobs while processing a validated certificate. This is not necessarily a block that is publishing a blob, it could just be a certificate using it.
So since it's a validated block, and it could be a block that's publishing one of these blobs, we need to check the chain manager, because that's the only place that blob might be at.
However, if it's just using the blob, it might already be in storage, so we check storage.

Reply:
Right, but with #2704 (if we go with what I suggested there) we'd only add the pending blobs to the chain manager when processing a _validated_ block (or a _fast_ block proposal), I think? So we'd look them up there when processing a _confirmed_ block?

And it probably doesn't hurt to look them up when processing a new proposal or a validated block either, if they are re-proposing a locked block from an earlier round.

(So maybe we don't actually need to change anything? I may just have suggested this earlier because I find it a bit confusing at the moment where we expect blobs to be a which stage of the process.)

Reply:
Even with the new way that the chain manager's blobs will be stored, we would still need a better distinction in the code about where the blobs can be in different stages of the process, no?
If we keep everything together like it was before and not change anything, the code IMHO is still confusing to figure out where the blobs can be 🤔
If we think that checking everywhere is actually the right way, and they _could_ be anywhere in the different stages of the process, then I'm probably misunderstanding something

Reply:
No, I agree, we need a better distinction. It's just that in this particular case it might make sense to look everywhere, after all?

Either way, I'd prefer the methods to be named after what they do, rather than where they are called.

Reply:
I'd still call this `check_no_missing_blocks`, so it's not named after the call site. Instead, I would add to the method documentation where it looks for the blobs. I actually wouldn't even mention a block there, since the method doesn't take a block as an argument.